### PR TITLE
feat: better mutant descriptions

### PR DESCRIPTION
**Pull request by nhaajt**
## Overview
+ Update mutator names, descriptions, and docs for better consistency
+ Add the ability to define better descriptions specific to each mutation
+ Add specific mutation descriptions for built-in mutators

## Specific mutation description
Specific mutation description can be defined per mutator that extends `TokenMutator` by overriding the `description` method, which provides the original and mutated string representation of the targeted token and the mutation location:
```scala
object CharClassNegation extends TokenMutator {
  override val name = "Character class negation"
  override val levels: Seq[Int] = Seq(1)
  override def description(original: String, mutated: String, location: Location): String =
    s"${location.pretty} Negate the character class `$original` to `$mutated`"

  override def mutate(token: RegexTree): Seq[Mutant] = ...
}
```

With the above definition, the new descriptions for a `CharClassNegation` would be:
```scala
// Original regex
"[ab0-9[A-Z][cd]]"
// Mutants
"[^ab0-9[A-Z][cd]]"
"[ab0-9[^A-Z][cd]]"
"[ab0-9[A-Z][^cd]]"

// Before (same for all mutants)
"Negate character class"
// After
"[0:0, 0:1) Negate the character class `[ab0-9[A-Z][cd]]` to `[^ab0-9[A-Z][cd]]`"
"[0:6, 0:7) Negate the character class `[A-Z]` to `[^A-Z]`"
"[0:11, 0:12) Negate the character class `[cd]` to `[^cd]`"
```

If more mutation information is desired, a mutation description can also be directly specified during mutation as an argument for the `toMutantOf`, `toMutantBeforeChildrenOf`, or `toMutantAfterChildrenOf` extension methods:
```scala
object CharClassChildRemoval extends TokenMutator {
  override val name: String = "Character class child removal"
  override val levels: Seq[Int] = Seq(2, 3)

  override def mutate(token: RegexTree): Seq[Mutant] = {
    def _mutate(token: Node): Seq[Mutant] = token.children map (child =>
      token
        .buildWhile(_ ne child)
        .toMutantOf(
          child,
          description =
            s"${child.location.pretty} Remove the child `${child.build}` from the character class `${token.build}`"
        )
    )

    token match {
      case cc: CharacterClass if cc.children.length > 1      => _mutate(cc)
      case cc: CharacterClassNaked if cc.children.length > 1 => _mutate(cc)
      case _                                                 => Nil
    }
  }
}
```

New mutant descriptions for a `CharClassChildRemoval` are shown below:
```scala
// Original regex
"[ab0-9[A-Z][cd]]"
// Mutants
"[b0-9[A-Z][cd]]"
"[a0-9[A-Z][cd]]"
"[ab[A-Z][cd]]"
"[ab0-9[cd]]"
"[ab0-9[A-Z]]"
"[ab0-9[A-Z][d]]"
"[ab0-9[A-Z][c]]"

// Before (same for all mutants)
"Remove a child from a character class"
// After
"[0:1, 0:2) Remove the child `a` from the character class `[ab0-9[A-Z][cd]]`"
"[0:2, 0:3) Remove the child `b` from the character class `[ab0-9[A-Z][cd]]`"
"[0:3, 0:6) Remove the child `0-9` from the character class `[ab0-9[A-Z][cd]]`"
"[0:6, 0:11) Remove the child `[A-Z]` from the character class `[ab0-9[A-Z][cd]]`"
"[0:11, 0:15) Remove the child `[cd]` from the character class `[ab0-9[A-Z][cd]]`"
"[0:12, 0:13) Remove the child `c` from the character class `[cd]`"
"[0:13, 0:14) Remove the child `d` from the character class `[cd]`"
```

## Breaking changes
- `TokenMutator.description` removed in favor of using `TokenMutator.description(original: String, mutated: String, location: Location): String` to define and generate a more detailed description for each mutant instead of the mutator.

### CODE REVIEW

File: core/src/main/scalajs/weaponregex/model/mutation/TokenMutatorJS.scala [hunk 1]
```diff
@@ -24,11 +24,6 @@ case class TokenMutatorJS(tokenMutator: TokenMutator) {
   @JSExport
   val levels: js.Array[Int] = tokenMutator.levels.toJSArray
 
-  /** A short description of the mutator
-    */
-  @JSExport
-  val description: String = tokenMutator.description
```
Comment (hugo-vrijswijk) [inline core/src/main/scalajs/weaponregex/model/mutation/TokenMutatorJS.scala:30]: Technically a breaking change, but I checked, and it's not used in stryker-js, so I think we can get away with it.

Could you also remove it from the `TokenMutator` interface in the `index.d.ts` file?

**Review by hugo-vrijswijk — COMMENTED** [review]
Nice! I like the implementation. It's pretty clean and makes sense. Some minor comments.

File: core/src/main/scala/weaponregex/model/Location.scala [hunk 2]
```diff
@@ -5,12 +5,14 @@ import scala.scalajs.js.annotation.*
...
-  *   start [[weaponregex.model.Position]]
+  *   start [[weaponregex.model.Position]] (inclusive)
   * @param end
-  *   end [[weaponregex.model.Position]]
+  *   end [[weaponregex.model.Position]] (exclusive)
   */
 @JSExportAll
-case class Location(start: Position, end: Position)
+case class Location(start: Position, end: Position) {
+  val pretty: String = s"[${start.pretty}, ${end.pretty})"
```
[the remark this answers is not in the document]
Comment (nhaajt) [reply core/src/main/scala/weaponregex/model/Location.scala]: Updated 🧐 Good to know.

File: core/src/main/scalajs/weaponregex/model/mutation/TokenMutatorJS.scala (hunk 1 above)
Comment (nhaajt) [reply core/src/main/scalajs/weaponregex/model/mutation/TokenMutatorJS.scala:30]: Perfect 👌 I'll update the PR later.

File: core/src/main/scala/weaponregex/model/mutation/TokenMutator.scala [hunk 3]
```diff
@@ -1,22 +1,39 @@
...
+  @deprecated(
+    "This member is deprecated and will be removed in a later version. " +
+      "For now, this will return the same value as `name`." +
+      "Description is now generated for each mutant instead of the mutator.",
+    "v1.2.0"
+  )
   /** A short description of the mutator
     */
-  def description: String
+  def description: String = name
```
[the remark this answers is not in the document]
Comment (nhaajt) [reply core/src/main/scala/weaponregex/model/mutation/TokenMutator.scala]: Done 👌

File: core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala [hunk 4]
```diff
@@ -44,25 +44,4 @@ object WeaponRegeXJS {
...
-    *   mutationLevels: [Target mutation levels. If this is `null`, the `mutators`, will not be filtered.],
-    *   flavor: [Regex flavor. By the default, `ParerFlavorJS` will be used.]
-    * }
-    *   }}}
-    * @return
-    *   A JavaScript Array of [[weaponregex.model.mutation.Mutant]] if can be parsed, or throw an exception otherwise
-    */
-  @deprecated("Use `mutate(pattern, flags, options)` instead. This will be removed in the future.", "0.7.x")
-  @JSExportTopLevel("mutate")
-  def mutate(pattern: String, options: MutationOptions = new MutationOptions): js.Array[MutantJS] =
```
Comment (hugo-vrijswijk) [inline core/src/main/scalajs/weaponregex/WeaponRegeXJS.scala:66]: Some of the tests still require these so those would need changes. You can revert this change as I've also removed it in #267

File: core/src/main/scala/weaponregex/model/mutation/TokenMutator.scala [hunk 5]
```diff
@@ -41,24 +58,24 @@ trait TokenMutator {
...
+      * @param description
+      *   The description of the mutant. If not provided, the default description of the mutator is used instead.
       * @return
       *   A [[weaponregex.model.mutation.Mutant]]
       */
-    def toMutantOf(token: RegexTree): Mutant = toMutantAt(token.location)
+    def toMutantOf(token: RegexTree, location: Location = null, description: String = null): Mutant = {
+      val loc: Location = if (location == null) token.location else location
+      val desc: String =
+        if (description == null) TokenMutator.this.description(token.build, pattern, loc) else description
```
[the remark this answers is not in the document]
Comment (nhaajt) [reply core/src/main/scala/weaponregex/model/mutation/TokenMutator.scala]: I initially used `Option` but for some reason, I switched to using `null`.
Anyhow, I switched it back to `Option`.

**Review by hugo-vrijswijk — APPROVED** [review]
LGTM feel free to (squash) merge
